src/manual/devel/best-practices/microservices.md

```diff
@@ -0,0 +1,307 @@
+
+To do so, install `eslint`, `babel-eslint`, `mocha-eslint`, `eslint-config-taskcluster`, and `eslint-plugin-taskcluster`.
+Add the following to the `scripts` section of `package.json`: `"test": "mocha .test/lint.js .test/*_test.js"`.
+
+Create `.eslintrc` in the root of the repository with
+```js
+{
+  "extends": "eslint-config-taskcluster"
+}
+```
+
+And the following in `test/lint.js`:
+```js
+var lint = require('mocha-eslint');
+
+var paths = [
+  'src/*.js',
+  'test/*.js',
+];
+
+lint(paths);
+```
+
+### Test Requirements
+
+A simple `git clone` .. `npm install` .. `npm test` should run successfully for new contributors.
+Anything else dramatically increases the difficulty in getting started.
+
+Where possible, try to write tests that do not require any credentials or access to external services.
+Liberal use of fakes, mocks, stubs, etc. allows most application logic to be tested in isolation.
+Note that `taskcluster-lib-loader` allows dependency injection by means of overwrites:
+
+```js
+let server = await load('server', {
+  profile: 'test',
+  dependency1: fakeDep1,
+  dependency2: fakeDep2,
+});
+```
+
+For tests that must have credentials, check for the presence of credentials and mark the suite as pending if they are not available:
+
+```js
+suite("things", function() {
+  if (!helper.haveRealCredentials) {
+    this.pending = true;
+  }
+});
+```
+
+This will generate clear output for anyone running the tests without credentials, showing that many tests were not run.
+If they make a pull request, then the full suite will run in automation, and any issues not detected locally will be revealed.
+
+## Taskcluster Libraries
+
+### General
+
+Do not use `taskcluster-lib-base`.
+Instead, depend directly on the `taskcluster-lib-*` libraries the service requires.
+
+The following sections describe best practices for specific platform libraries.
+
+### taskcluster-lib-loader
+
+The main entry-point for the service should be `src/main.js`, which should use [taskcluster-lib-loader](https://github.com/taskcluster/taskcluster-lib-loader) and have the following initialization code:
+
+```js
+// If this file is executed launch component from first argument
+if (!module.parent) {
+  load(process.argv[2], {
+    process: process.argv[2],
+    profile: process.env.NODE_ENV,
+  }).catch(err => {
+    console.log(err.stack);
+    process.exit(1);
+  });
+}
+
+// Export load for tests
+module.exports = load;
+```
+
+Entries in `Procfile`, then, look like `web: node lib/main.js server`.
+The web service should always be the component named `server`.
+All services, including those run from the Heroku scheduler, should start like this, via `lib/main.js`.
+
+### azure-entities
+
+Each Azure table should be defined in `src/data.js` using a cascade of `configure` calls, one for each version:
```

Comment:
That would be a good additional PR-and-cleanup.  A grep shows most services are using data.js.